pkg/quota/controller/clusterquotareconciliation/reconcilation_controller.go

```diff
@@ -0,0 +1,357 @@
+	obj, err := kapi.Scheme.Copy(originalQuota)
+	if err != nil {
+		return err, workItems
+	}
+	quota := obj.(*quotaapi.ClusterResourceQuota)
+
+	// get the list of namespaces that match this cluster quota
+	matchingNamespaceNamesList, quotaSelector := c.clusterQuotaMapper.GetNamespacesFor(quota.Name)
+	if !kapi.Semantic.DeepEqual(quotaSelector, quota.Spec.Selector) {
+		return fmt.Errorf("mapping not up to date, have=%v need=%v", quotaSelector, quota.Spec.Selector), workItems
+	}
+	matchingNamespaceNames := sets.NewString(matchingNamespaceNamesList...)
+
+	reconcilationErrors := []error{}
+	retryItems := []workItem{}
+	for _, item := range workItems {
+		namespaceName := item.namespaceName
+		namespaceTotals, namespaceLoaded := quota.Status.Namespaces.Get(namespaceName)
+		if !matchingNamespaceNames.Has(namespaceName) {
+			if namespaceLoaded {
+				// remove this item from all totals
+				quota.Status.Total.Used = utilquota.Subtract(quota.Status.Total.Used, namespaceTotals.Used)
+				quota.Status.Namespaces.Remove(namespaceName)
+			}
+			continue
+		}
+
+		// if there's no work for us to do, do nothing
+		if !item.forceRecalculation && namespaceLoaded && kapi.Semantic.DeepEqual(namespaceTotals.Hard, quota.Spec.Quota.Hard) {
+			continue
+		}
+
+		actualUsage, err := quotaUsageCalculationFunc(namespaceName, quota.Spec.Quota.Scopes, quota.Spec.Quota.Hard, c.registry)
+		if err != nil {
+			// tally up errors, but calculate everything you can
+			reconcilationErrors = append(reconcilationErrors, err)
+			retryItems = append(retryItems, item)
+			continue
+		}
+		recalculatedStatus := kapi.ResourceQuotaStatus{
+			Used: actualUsage,
+			Hard: quota.Spec.Quota.Hard,
+		}
+
+		// subtract old usage, add new usage
+		quota.Status.Total.Used = utilquota.Subtract(quota.Status.Total.Used, namespaceTotals.Used)
+		quota.Status.Total.Used = utilquota.Add(quota.Status.Total.Used, recalculatedStatus.Used)
+		quota.Status.Namespaces.Insert(namespaceName, recalculatedStatus)
+	}
+
+	quota.Status.Total.Hard = quota.Spec.Quota.Hard
+
+	// if there's no change, no update, return early.  NewAggregate returns nil on empty input
+	if kapi.Semantic.DeepEqual(quota, originalQuota) {
+		return kutilerrors.NewAggregate(reconcilationErrors), retryItems
+	}
+
+	// TODO separate out status updating
+	if _, err := c.clusterQuotaClient.ClusterResourceQuotas().Update(quota); err != nil {
+		return kutilerrors.NewAggregate(append(reconcilationErrors, err)), workItems
+	}
+
+	return kutilerrors.NewAggregate(reconcilationErrors), retryItems
+}
+
+// replenishQuota is a replenishment function invoked by a controller to notify that a quota should be recalculated
+func (c *ClusterQuotaReconcilationController) replenishQuota(groupKind unversioned.GroupKind, namespace string, object runtime.Object) {
+	// check if the quota controller can evaluate this kind, if not, ignore it altogether...
+	evaluators := c.registry.Evaluators()
+	evaluator, found := evaluators[groupKind]
```

Comment:
a lot of the function is a clone of upstream.

i don't have time to think too deeply on this now, but it would be good if we can structure the upstream call a little differently so we can not diverge on structure especially when we look to tackle targeted replenishment per namespace for a queued list of kinds.  although, maybe all the shared informer work will lessen the need for that.